### PR TITLE
REGR: read_pickle fallback to encoding=latin_1 upon a UnicodeDecodeError

### LINKED ISSUE #31988: Error in `read_pickle` when loading a DataFrame with MultiIndex columns from a pickle created in py27
Calling `read_pickle` to load a DataFrame with MultiIndex columns from a pickle file created in py27 throws an error saying:
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xd9 in position 0: ordinal not in range(128)

This issue was introduced by #28645, more specifically by [this change](https://github.com/pandas-dev/pandas/pull/28645/files#diff-7ecbc1597b1273ca207747232d64edad):

```diff
diff --git a/pandas/io/pickle.py b/pandas/io/pickle.py
index adf0aa961..8f9bae0f7 100644
--- a/pandas/io/pickle.py
+++ b/pandas/io/pickle.py
@@ -142,18 +142,24 @@ def read_pickle(path, compression="infer"):

     # 1) try standard library Pickle
     # 2) try pickle_compat (older pandas version) to handle subclass changes
-    # 3) try pickle_compat with latin1 encoding
+
+    excs_to_catch = (AttributeError, ImportError)
+    if PY36:
+        excs_to_catch += (ModuleNotFoundError,)

     try:
         with warnings.catch_warnings(record=True):
             # We want to silence any warnings about, e.g. moved modules.
             warnings.simplefilter("ignore", Warning)
             return pickle.load(f)
-    except Exception:
-        try:
-            return pc.load(f, encoding=None)
-        except Exception:
-            return pc.load(f, encoding="latin1")
+    except excs_to_catch:
+        # e.g.
+        #  "No module named 'pandas.core.sparse.series'"
+        #  "Can't get attribute '__nat_unpickle' on <module 'pandas._libs.tslib"
+        return pc.load(f, encoding=None)
+    except UnicodeDecodeError:
+        # e.g. can occur for files written in py27; see GH#28645
+        return pc.load(f, encoding="latin-1")
     finally:
         f.close()
         for _f in fh:
```

Note how before when there was an Exception trying to load the file with the builtin `pickle.load` function, it would try to load them using `pickle_compat` with `enconding=None` and if that call also threw an Exception, it would then fallback to try to load with `encoding="latin-1"`. With the change from #28645 the fallback to use `encoding="latin-1"` is only in the catch block of the initial `pickle_load` call, not the second one that tries to use pickle_compat with `encoding=None`.

This become an issue for py27 pickles with MultiIndex columns after `FrozenNDArray` was removed by #29840 as `pickle.load(f)` throws an `AttributeError` for  `FrozenNDArray` and then `pc.load(f, encoding=None)` throws an `UnicodeDecodeError`.

Here is a full stack trace:
<details>

```python
In [1]: import pandas as pd
   ...: df = pd.read_pickle('test_mi_py27.pkl')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/anaconda3/envs/pandas-pickle/lib/python3.7/site-packages/pandas/io/pickle.py in read_pickle(filepath_or_buffer, compression)
    180             warnings.simplefilter("ignore", Warning)
--> 181             return pickle.load(f)
    182     except excs_to_catch:

AttributeError: Can't get attribute 'FrozenNDArray' on <module 'pandas.core.indexes.frozen' from '~/anaconda3/envs/pandas-pickle/lib/python3.7/site-packages/pandas/core/indexes/frozen.py'>

During handling of the above exception, another exception occurred:

UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-1-8c3800ea96cb> in <module>
      1 import pandas as pd
----> 2 df = pd.read_pickle('test_mi_py27.pkl')

~/anaconda3/envs/pandas-pickle/lib/python3.7/site-packages/pandas/io/pickle.py in read_pickle(filepath_or_buffer, compression)
    184         #  "No module named 'pandas.core.sparse.series'"
    185         #  "Can't get attribute '__nat_unpickle' on <module 'pandas._libs.tslib"
--> 186         return pc.load(f, encoding=None)
    187     except UnicodeDecodeError:
    188         # e.g. can occur for files written in py27; see GH#28645

~/anaconda3/envs/pandas-pickle/lib/python3.7/site-packages/pandas/compat/pickle_compat.py in load(fh, encoding, is_verbose)
    239         up.is_verbose = is_verbose
    240
--> 241         return up.load()
    242     except (ValueError, TypeError):
    243         raise

~/anaconda3/envs/pandas-pickle/lib/python3.7/pickle.py in load(self)
   1086                     raise EOFError
   1087                 assert isinstance(key, bytes_types)
-> 1088                 dispatch[key[0]](self)
   1089         except _Stop as stopinst:
   1090             return stopinst.value

~/anaconda3/envs/pandas-pickle/lib/python3.7/pickle.py in load_short_binstring(self)
   1262         len = self.read(1)[0]
   1263         data = self.read(len)
-> 1264         self.append(self._decode_string(data))
   1265     dispatch[SHORT_BINSTRING[0]] = load_short_binstring
   1266

~/anaconda3/envs/pandas-pickle/lib/python3.7/pickle.py in _decode_string(self, value)
   1202             return value
   1203         else:
-> 1204             return value.decode(self.encoding, self.errors)
   1205
   1206     def load_string(self):

UnicodeDecodeError: 'ascii' codec can't decode byte 0xd9 in position 0: ordinal not in range(128)
```

</details>

I believe the fix for this specific issue can be as simple as:

```diff
diff --git a/pandas/io/pickle.py b/pandas/io/pickle.py
index e51f24b55..5c4f2d8c4 100644
--- a/pandas/io/pickle.py
+++ b/pandas/io/pickle.py
@@ -183,7 +183,11 @@ def read_pickle(
         # e.g.
         #  "No module named 'pandas.core.sparse.series'"
         #  "Can't get attribute '__nat_unpickle' on <module 'pandas._libs.tslib"
-        return pc.load(f, encoding=None)
+        try:
+            return pc.load(f, encoding=None)
+        except UnicodeDecodeError:
+            # e.g. can occur for files written in py27;
+            return pc.load(f, encoding="latin-1")
     except UnicodeDecodeError:
         # e.g. can occur for files written in py27; see GH#28645
         return pc.load(f, encoding="latin-1")
```

I will open a PR with the fix above over the weekend.

Thanks!

### PR DESCRIPTION
**Pull request by pedroreys**
When a reading a pickle with MultiIndex columns generated in py27
`pickle_compat.load()` with `enconding=None` would throw an UnicodeDecodeError
when reading a pickle created in py27. Now, `read_pickle` catches that exception and
fallback to use `latin-1` explicitly.

- [x] closes #31988
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/pickle.py [hunk 1]
```diff
@@ -183,7 +183,11 @@ def read_pickle(
...
         #  "No module named 'pandas.core.sparse.series'"
         #  "Can't get attribute '__nat_unpickle' on <module 'pandas._libs.tslib"
-        return pc.load(f, encoding=None)
+        try:
+            return pc.load(f, encoding=None)
+        except UnicodeDecodeError:
+            # e.g. can occur for files written in py27; see GH#28645 and GH#31988
+            return pc.load(f, encoding="latin-1")
     except UnicodeDecodeError:
         # e.g. can occur for files written in py27; see GH#28645
```
Comment (jbrockmendel) [inline pandas/io/pickle.py]: is there any way we can avoid catching UnicodeDecodeError twice?

Comment (pedroreys) [reply pandas/io/pickle.py]: updated so that it only has a single catch for UnicodeDecodeError

File: pandas/tests/io/test_pickle.py [hunk 2]
```diff
@@ -501,3 +501,15 @@ def test_read_pickle_with_subclass():
 
     tm.assert_series_equal(result[0], expected[0])
     assert isinstance(result[1], MyTz)
+
+
+def test_read_py27_pickle_with_MultiIndex_column(datapath):
```
Comment (simonjayhawkins) [inline pandas/tests/io/test_pickle.py]: can this test be moved to just after `test_unicode_decode_error`, better still parameterize `test_unicode_decode_error` with `filename` and `excols`

File: doc/source/whatsnew/v1.0.2.rst [hunk 3]
```diff
@@ -36,6 +36,7 @@ Bug fixes
 
 - Using ``pd.NA`` with :meth:`DataFrame.to_json` now correctly outputs a null value instead of an empty object (:issue:`31615`)
 - Fixed bug in parquet roundtrip with nullable unsigned integer dtypes (:issue:`31896`).
+- Fixed bug where :meth:`pandas.io.pickleread_pickle` raised a ``UnicodeDecodeError` when reading a py27 pickle with MultiIndex column (:issue:`31988`).
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.0.2.rst]:
```suggestion
- Fixed regression where :func:`read_pickle` raised a ``UnicodeDecodeError`` when reading a py27 pickle with :class:`MultiIndex` column (:issue:`31988`).
```

and move to regression section above.

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @pedroreys generally looks good. couple of suggestions.

File: pandas/tests/io/test_pickle.py (hunk 2 above)
Comment (pedroreys) [reply pandas/tests/io/test_pickle.py]: done

File: doc/source/whatsnew/v1.0.2.rst [hunk 4]
```diff
@@ -53,4 +55,4 @@ Bug fixes
 Contributors
 ~~~~~~~~~~~~
 
-.. contributors:: v1.0.1..v1.0.2|HEAD
\ No newline at end of file
+.. contributors:: v1.0.1..v1.0.2|HEAD
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.2.rst]: Is this modified here for a reason? If not can you revert?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks good; question on the whastnew change

**Review by simonjayhawkins — APPROVED** [review]
@pedroreys thanks for the changes. lgtm ex @WillAyd comment

File: doc/source/whatsnew/v1.0.2.rst (hunk 4 above)
Comment (pedroreys) [reply doc/source/whatsnew/v1.0.2.rst]: That was not on purpose, I will revert.

Comment (pedroreys) [reply doc/source/whatsnew/v1.0.2.rst]: @WillAyd done. Thank you for the feedback.

File: pandas/io/pickle.py [hunk 5]
```diff
@@ -171,21 +171,22 @@ def read_pickle(
...
-        with warnings.catch_warnings(record=True):
-            # We want to silence any warnings about, e.g. moved modules.
-            warnings.simplefilter("ignore", Warning)
-            return pickle.load(f)
-    except excs_to_catch:
-        # e.g.
-        #  "No module named 'pandas.core.sparse.series'"
-        #  "Can't get attribute '__nat_unpickle' on <module 'pandas._libs.tslib"
-        return pc.load(f, encoding=None)
+        excs_to_catch = (AttributeError, ImportError, ModuleNotFoundError)
```
Comment (jbrockmendel) [inline pandas/io/pickle.py]: is there a reason to move this inside the "try"?

Comment (pedroreys) [reply pandas/io/pickle.py]: actually, I didn’t move this to inside the try. It is still outside the same`try` as it was before (semantically).

What I did in order to catch `UnicodeDecodeError` only once was to wrap all that code in an outer `try` block that catches `UnicodeDecodeError` regardless if is thrown by `pickle.load(f)` or by `pc.load(f, encoding=none)`.

So, in summary, it’s not that I moved the `excs_to_catch` to inside the `try`, but that I added a new `try` block just before it.

Comment (jbrockmendel) [reply pandas/io/pickle.py]: ok thanks

**Comment by jbrockmendel** [comment]
LGTM

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @pedroreys !

**Pull request merged**